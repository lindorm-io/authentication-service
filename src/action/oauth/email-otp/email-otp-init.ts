import Joi from "@hapi/joi";
import { IAuthContext } from "../../../typing";
import { JOI_CODE_CHALLENGE, JOI_CODE_METHOD, JOI_EMAIL, JOI_GRANT_TYPE, JOI_STATE } from "../../../constant";
import { assertValidScopeInput, assertValidResponseTypeInput } from "../../../util";
import { createSession, getAuthorizationToken, sendEmailOTP } from "../../../support";
import { getRandomNumber } from "@lindorm-io/core";

export interface IPerformEmailOTPInitOptions {
  codeChallenge: string;
  codeMethod: string;
  grantType: string;
  redirectUri: string;
  responseType: string;
  scope: string;
  state: string;
  subject: string;
}

export interface IPerformEmailOTPInitData {
  expires: number;
  expiresIn: number;
  redirectUri: string;
  state: string;
  token: string;
}

const schema = Joi.object({
  codeChallenge: JOI_CODE_CHALLENGE,
  codeMethod: JOI_CODE_METHOD,
  grantType: JOI_GRANT_TYPE,
  redirectUri: Joi.string().uri().required(),
  responseType: Joi.string().required(),
  scope: Joi.string().required(),
  state: JOI_STATE,
  subject: JOI_EMAIL,
});

export const performEmailOTPInit = (ctx: IAuthContext) => async (
  options: IPerformEmailOTPInitOptions,
): Promise<IPerformEmailOTPInitData> => {
  await schema.validateAsync(options);

  const { client } = ctx;
  const { codeChallenge, codeMethod, grantType, redirectUri, responseType, scope, state, subject } = options;

  assertValidResponseTypeInput(responseType);
  assertValidScopeInput(scope);

  const otpCode = await getRandomNumber(6);

  const session = await createSession(ctx)({
    codeChallenge,
    codeMethod,
    grantType,
    otpCode,
    redirectUri,
    responseType,
    scope,
    state,
    subject,
  });

  const { expires, expiresIn, token } = getAuthorizationToken(ctx)({ client, session });

  await sendEmailOTP(ctx)({ otpCode, subject });

  return {
    expires,
    expiresIn,
    redirectUri,
    state,
    token,
  };
};
