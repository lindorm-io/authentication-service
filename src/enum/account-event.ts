export enum AccountEvent {
  CREATED = "account_created",
  EMAIL_CHANGED = "account_email_changed",
  OTP_CHANGED = "account_otp_changed",
  PASSWORD_CHANGED = "account_password_changed",
  PERMISSION_CHANGED = "account_permission_changed",
}
