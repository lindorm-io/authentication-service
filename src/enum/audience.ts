export enum Audience {
  ACCESS = "access",
  AUTHORIZATION = "authorization",
  IDENTITY = "identity",
  MULTI_FACTOR = "multi-factor",
  REFRESH = "refresh",
}
