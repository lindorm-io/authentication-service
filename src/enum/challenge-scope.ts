export enum ChallengeScope {
  CHANGE = "change",
  SIGN_IN = "sign_in",
  RECOVERY = "recovery",
}
