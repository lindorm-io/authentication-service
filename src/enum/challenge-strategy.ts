export enum ChallengeStrategy {
  IMPLICIT = "implicit",
  PIN = "pin",
  SECRET = "secret",
}
