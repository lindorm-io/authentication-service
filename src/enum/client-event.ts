export enum ClientEvent {
  CREATED = "client_created",
  APPROVED_CHANGED = "client_approved_changed",
  DESCRIPTION_CHANGED = "client_description_changed",
  EMAIL_AUTHORIZATION_URI_CHANGED = "client_email_authorization_uri_changed",
  NAME_CHANGED = "client_name_changed",
  SECRET_CHANGED = "client_secret_changed",
}
