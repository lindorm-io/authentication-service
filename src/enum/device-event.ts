export enum DeviceEvent {
  CREATED = "device_created",
  PIN_CHANGED = "device_pin_changed",
  SECRET_CHANGED = "device_secret_changed",
  NAME_CHANGED = "device_name_changed",
}
