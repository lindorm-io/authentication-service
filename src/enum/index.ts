export * from "./account-event";
export * from "./audience";
export * from "./device-event";
export * from "./grant-type";
export * from "./mongo-collection";
export * from "./multi-factor-challenge-type";
export * from "./response-type";
export * from "./session-event";
