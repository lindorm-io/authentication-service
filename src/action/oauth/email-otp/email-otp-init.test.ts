import { GrantType, ResponseType } from "../../../enum";
import { Scope } from "@lindorm-io/jwt";
import { performEmailOTPInit } from "./email-otp-init";

jest.mock("../../../util", () => ({
  assertValidResponseTypeInput: jest.fn(),
  assertValidScopeInput: jest.fn(),
}));
jest.mock("../../../support", () => ({
  createAuthorization: jest.fn(() => () => "createAuthorization"),
  getAuthorizationToken: jest.fn(() => () => ({
    expires: "expires",
    expiresIn: "expiresIn",
    token: "token",
  })),
  sendEmailOTP: jest.fn(() => () => {}),
}));
jest.mock("@lindorm-io/core", () => ({
  getRandomNumber: jest.fn(() => "123456"),
}));

describe("performEmailOTPInit", () => {
  let ctx: any;

  beforeEach(() => {
    ctx = {
      client: "client",
    };
  });

  test("should create a new session", async () => {
    await expect(
      performEmailOTPInit(ctx)({
        codeChallenge: "Z1teIWMlf6xFacp4quXP3O0XI204ZT1b",
        codeMethod: "sha512",
        grantType: GrantType.DEVICE_PIN,
        redirectUri: "https://redirect.uri/",
        responseType: ResponseType.REFRESH,
        scope: Scope.DEFAULT,
        state: "1bVcJqZ1pBeqVLxV",
        subject: "email@lindorm.io",
      }),
    ).resolves.toMatchSnapshot();
  });
});
