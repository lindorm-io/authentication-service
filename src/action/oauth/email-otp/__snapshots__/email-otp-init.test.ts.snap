// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`performEmailOTPInit should create a new session 1`] = `
Object {
  "expires": "expires",
  "expiresIn": "expiresIn",
  "redirectUri": "https://redirect.uri/",
  "state": "1bVcJqZ1pBeqVLxV",
  "token": "token",
}
`;
